python_modules/dagster/dagster/core/files.py

```diff
@@ -0,0 +1,61 @@
+from abc import ABCMeta, abstractmethod
+from contextlib import contextmanager
+import os
+
+import six
+
+from dagster import check
+from dagster.utils import mkdir_p
+
+
+class FileStore(six.with_metaclass(ABCMeta)):  # pylint: disable=no-init
+    @abstractmethod
+    @contextmanager
+    def writeable_binary_stream(self, *path_comps):
+        pass
+
+    @abstractmethod
+    @contextmanager
+    def readable_binary_stream(self, *path_comps):
+        pass
+
+
+class LocalTempFileStore(FileStore):
+    def __init__(self, run_id):
+        check.str_param(run_id, 'run_id')
+        self.root = os.path.join('/tmp', 'dagster', 'runs', run_id, 'files')
+        mkdir_p(self.root)
+
+    @contextmanager
+    def writeable_binary_stream(self, *path_comps):
```

Comment:
in the S3 case, do you have in mind that we would buffer this stream somewhere and then upload it to s3 when we yield back to the context manager, or do a streaming multipart upload and never hold it locally?

Reply:
Hmmm i might not be understanding the question. I was thinking the s3 implementation would just yield an object the implements the write method, and then it does whatever it needs to do in order to write and flush it to s3. (We also might need to use emrfs for its consistency guarantees, different subject).